app/controllers/incomeSources/cease/CeaseIncomeSourceController.scala

```diff
@@ -16,31 +16,87 @@
-                                            val checkSessionTimeout: SessionTimeoutPredicate)
-                                           (implicit val appConfig: FrontendAppConfig,
-                                            mcc: MessagesControllerComponents,
-                                            val ec: ExecutionContext,
-                                            val itvcErrorHandlerAgent: AgentItvcErrorHandler)
+class CeaseIncomeSourceController @Inject()(val ceaseIncomeSources: CeaseIncomeSources,
+                                            val checkSessionTimeout: SessionTimeoutPredicate,
+                                            val authenticate: AuthenticationPredicate,
+                                            val authorisedFunctions: AuthorisedFunctions,
+                                            val retrieveNino: NinoPredicate,
+                                            val retrieveIncomeSources: IncomeSourceDetailsPredicate,
+                                            val itvcErrorHandler: ItvcErrorHandler,
+                                            implicit val itvcErrorHandlerAgent: AgentItvcErrorHandler,
+                                            val incomeSourceDetailsService: IncomeSourceDetailsService,
+                                            val retrieveBtaNavBar: NavBarPredicate)
+                                           (implicit val ec: ExecutionContext,
+                                            implicit override val mcc: MessagesControllerComponents,
+                                            val appConfig: FrontendAppConfig)
   extends ClientConfirmedController with FeatureSwitching with I18nSupport {
 
-  def show(origin: Option[String] = None): Action[AnyContent] = Action {
-    Ok("")
+  def show(): Action[AnyContent] = (checkSessionTimeout andThen authenticate andThen retrieveNino
+    andThen retrieveIncomeSources andThen retrieveBtaNavBar).async {
+    implicit user =>
+      handleRequest(
+        sources = user.incomeSources,
+        isAgent = false,
+        backUrl = controllers.routes.HomeController.show().url
+    )
   }
 
-  def showAgent(): Action[AnyContent] = Action {
-    Ok("")
+  def showAgent(): Action[AnyContent] = Authenticated.async {
+    implicit request =>
+      implicit user =>
+        getMtdItUserWithIncomeSources(incomeSourceDetailsService) flatMap {
+          implicit mtdItUser =>
+            handleRequest(
+              sources = mtdItUser.incomeSources,
+              isAgent = true,
+              backUrl = controllers.routes.HomeController.showAgent.url
+            )
+        }
   }
-}
+
+  def handleRequest(sources: IncomeSourceDetailsModel, isAgent: Boolean, backUrl: String)
+                   (implicit user: MtdItUser[_]): Future[Result] = {
+    Future.successful(
+      if (isDisabled(IncomeSources)) {
+        Redirect(controllers.routes.HomeController.show())
```

Comment:
* pls wrap redirect with: Future.successful : which is completed Future value;
* for the incomeSourceDetailsService.getCeaseIncomeSourceViewModel call below, pls use actual Future, as this would be subject to evaluation etc

Reply:
I would believe you should wrap incomeSourceDetailsService.getCeaseIncomeSourceViewModel into the Future as well; you can think that IF i.e getCeaseIncomeSourceViewModel would return Future after some rework/or possible changes, this would really require this whole section to be Future

Reply:
Added future!